### DISCUSSION TITLE
Varlen attention vs multipack masking

### DISCUSSION THREAD

**Discussion by euclaise** [axolotl-ai-cloud/axolotl#920] [Ideas]
At the current ollama event, multipack was mentioned as using masking to prevent cross-sequence attention. 

However, it seems to me like this needlessly wastes memory - since attention memory/compute is quadratic with seq len, it should be more efficient to run N attentions batched in parallel. 

It seems that flash-attn already supports variable-length batching, and its varlen functions are even used in the patch scripts. 

Is there a reason to prefer masking that I'm missing?

**Comment by NanoCode012** [comment]
Extremely late reply but thought I'd give one nevertheless for any curious readers.

The multipack in Axolotl uses attention varlen (originally FA's) to pack multiple samples together to reduce padding. This is useful when targeting a higher context length as there won't be as much padding tokens if it's filled with actual tokens. Additionally, we use position_ids to prevent cross-contamination during attn.

Therefore, it's faster as there's less total number of samples while still training on much longer context.
